### PR TITLE
Loading stategy compat

### PR DESCRIPTION
**Pull request by peterwilsoncc**
Follow up to:

* https://core.trac.wordpress.org/ticket/59316
* https://core.trac.wordpress.org/ticket/59302

### CODE REVIEW

File: src/wp-content/themes/twentyfourteen/functions.php [hunk 1]
```diff
@@ -372,11 +378,19 @@ function twentyfourteen_scripts() {
 			get_template_directory_uri() . '/js/slider.js',
 			array( 'jquery' ),
 			'20150120',
-			array(
-				'in_footer' => false, // Because involves header.
-				'strategy'  => 'defer',
-			)
+			true
```
Comment (felixarntz) [inline src/wp-content/themes/twentyfourteen/functions.php:381]: This change in particular may make performance worse on WP 6.3+: The slider appears in the header, so with `defer` it should also be placed in the <head> to be available early.

File: src/wp-content/themes/twentynineteen/functions.php [hunk 2]
```diff
@@ -263,21 +266,27 @@ function twentynineteen_scripts() {
 			get_theme_file_uri( '/js/priority-menu.js' ),
 			array(),
 			'20200129',
-			array(
-				'in_footer' => false, // Because involves header.
-				'strategy'  => 'defer',
-			)
+			true
```
Comment (felixarntz) [inline src/wp-content/themes/twentynineteen/functions.php:269]: See above.

File: src/wp-content/themes/twentyseventeen/functions.php [hunk 3]
```diff
@@ -477,18 +480,25 @@ function twentyseventeen_scripts() {
...
 	wp_enqueue_script(
 		'twentyseventeen-global',
 		get_theme_file_uri( '/assets/js/global.js' ),
 		array( 'jquery' ),
 		'20211130',
-		array(
-			'in_footer' => false, // Because involves header.
-			'strategy'  => 'defer',
-		)
+		true
```
Comment (felixarntz) [inline src/wp-content/themes/twentyseventeen/functions.php:490]: See above (also applies to a few more places in the PR).

**Review by felixarntz — CHANGES_REQUESTED** [review]
Reviewing the original commits, I now see part of what this change is trying to achieve makes sense. However, the way this PR implements them does not make sense to me. It is effectively a revert of most of the work from https://core.trac.wordpress.org/ticket/59302 and https://core.trac.wordpress.org/ticket/59316, and I can't see why.

Specifically:
* _In general_ using the new function signature does not break backward compatibility. A non-empty array will be interpreted as `true` in pre-WP 6.3, which is okay since in all cases where a non-empty array is passed, the asset could in older versions be printed in the footer. The exception to that is assets which should in older versions be printed in the head, in which case the parameter should simply be omitted.
* When I asked before on the ticket, in https://core.trac.wordpress.org/ticket/59302#comment:5 you mention "choosing to pass around the incorrect type doesn't seem like a great idea", yet this PR does exactly the same, just using the old signature instead of the new one, which is not better or worse.

Now to the actual problems with the committed code that I can see:
* https://core.trac.wordpress.org/changeset/56526 does not cause any issues and is fully backward compatible.
* In https://core.trac.wordpress.org/changeset/56556 however, there are indeed a few places that are not backward compatible because in pre-WP 6.3 their script location would now have changed from what it was before. Those are specifically:
    * https://core.trac.wordpress.org/changeset/56556/trunk/src/wp-content/themes/twentyeleven/showcase.php
    * https://core.trac.wordpress.org/changeset/56556/trunk/src/wp-content/themes/twentytwenty/functions.php

In other words, those two instances are actual problems and should be fixed. For the rest of the changes in this PR, I don't see a good reason.

**Comment by westonruter** [comment]
> * In https://core.trac.wordpress.org/changeset/56556 however, there are indeed a few places that are not backward compatible because in pre-WP 6.3 their script location would now have changed from what it was before. Those are specifically:
>   
>   * https://core.trac.wordpress.org/changeset/56556/trunk/src/wp-content/themes/twentyeleven/showcase.php

I just noticed this as well, but I explained how this is not a backwards-compatibility break in https://github.com/WordPress/wordpress-develop/pull/5170/files#r1324822884

For the showcase script, this actually was originally added to the `head`, so in WP<6.3 this will actually get printed in the footer since the `$args` array is truthy for `$in_footer`. Nevertheless, the function's contents are wrapped in `jQuery.ready()` so it makes no difference when it loads.

> https://core.trac.wordpress.org/changeset/56556/trunk/src/wp-content/themes/twentytwenty/functions.php

In the same way, the `index.js` script for Twenty Twenty is designed to be loaded asynchronously, so it doesn't matter where the script loads, whether it is in the `head` or footer.

**Comment by felixarntz** [comment]
@westonruter I still consider those two changes backward compatibility breaks because the code does not work the same way as it did before, and I don't think there's a good reason to move these two scripts to the footer? They were in the head originally, and whether that decision was originally made with caution or not, I think the scripts affect content in the head, so we shouldn't just move them to the footer in older versions, specifically for Twenty Eleven where there's no shim for async/defer. Even for Twenty Twenty though, you mention `async`, but the loading strategy was actually changed in the new PR to `defer`. I know that one was intentional, but for older versions I think it still results in an unintentional change of `head async` to `footer defer`.

**Comment by westonruter** [comment]
> @westonruter I still consider those two changes backward compatibility breaks because the code does not work the same way as it did before, and I don't think there's a good reason to move these two scripts to the footer? They were in the head originally, and whether that decision was originally made with caution or not, I think the scripts affect content in the head, so we shouldn't just move them to the footer in older versions, specifically for Twenty Eleven where there's no shim for async/defer. 

@felixarntz While the script location does move, there is no functional backwards-compatibility break because the logic is wrapped in `jQuery.ready()`. In fact, there is actually a good reason to move it to the footer because it prevents a render-blocking script form being needlessly in the `head`. In WP 6.3, the script is kept in the `head` because it can get `defer`. But in WP<6.3 the side effect of passing an array for `$in_footer` is that it gets moved to the footer, which is also good for performance. The lack of the `$in_footer` parameter in Twenty Eleven may have been an oversight or due to the `$in_footer` parameter being "relatively" new to WordPress core. 

> Even for Twenty Twenty though, you mention `async`, but the loading strategy was actually changed in the new PR to `defer`. I know that one was intentional, but for older versions I think it still results in an unintentional change of `head async` to `footer defer`.

Actually, I meant it was designed to load _asynchronously_ not necessarily `async`. In other words, it is designed to load before or after `DOMContentLoaded`. But yes, in this case the script doesn't have a benefit of being moved to the footer and that should be revised with the `wp_script_add_data()` workaround.

**Update:** I've opened https://github.com/WordPress/wordpress-develop/pull/5204 to address this.

**Comment by felixarntz** [comment]
Thanks for clarifying @westonruter. Both of those explanations make sense to me. Though one follow up question:
> The lack of the `$in_footer` parameter in Twenty Eleven may have been an oversight or due to the `$in_footer` parameter being "relatively" new to WordPress core. 

This is speaking from a performance perspective, but what about the functionality that the script covers? Could it be considered essential enough to have the script in the head? Speaking hypothetically, if `async` and `defer` did not exist, there would probably still be some scripts that should be in the head if they cover critical content, even if it's worse for load time performance, that's where my question is coming from.

**Comment by westonruter** [comment]
> This is speaking from a performance perspective, but what about the functionality that the script covers? Could it be considered essential enough to have the script in the head? Speaking hypothetically, if `async` and `defer` did not exist, there would probably still be some scripts that should be in the head if they cover critical content, even if it's worse for load time performance, that's where my question is coming from.

@felixarntz You mean the `showcase.js` script? The [logic](https://github.com/WordPress/wordpress-develop/blob/6f74f616b8973fdef00e81e0984d33c31b229371/src/wp-content/themes/twentyeleven/js/showcase.js#L2-L16) is wrapped in `jQuery.ready()` (aka `$()`) so it is deferred to execute until `DOMContentLoaded` anyway. So there is no functional difference. This `DOMContentLoaded` event is necessary for any such scripts loading in the `head` because without it none of the `body` would be loaded; so here the showcase script's `$('.feature-slider a')` query would return no results. The only way to work around the use of a `DOMContentLoaded` event handler is to rely on event delegation or MutationEvents, neither of which Twenty Eleven is using. So yeah, functionally the showcase script can be loaded either at `wp_head` or `wp_footer` without any difference in functionality.

**Comment by peterwilsoncc** [comment]
@westonruter @felixarntz Thanks for hunting this out after I forgot to request reviews once the tests passed. I'll follow it up later today.

**Comment by peterwilsoncc** [comment]
> * When I asked before on the ticket, in https://core.trac.wordpress.org/ticket/59302#comment:5 you mention "choosing to pass around the incorrect type doesn't seem like a great idea", yet this PR does exactly the same, just using the old signature instead of the new one, which is not better or worse.

There's a significant difference here: in 6.3+ there is a type check for the `$args` parameter that doesn't exist for the `$in_footer` argument in earlier versions of WP.

A part of the motivation with these changes is to encourage future additions to these themes to use the legacy signature when supporting new features in WP (most likely the block editor). `[ `in_footer` => false ]` will load scripts in the header in older versions of WP.

File: src/wp-content/themes/twentyfourteen/functions.php (hunk 1 above)
Comment (peterwilsoncc) [reply src/wp-content/themes/twentyfourteen/functions.php:381]: > This change in particular may make performance worse on WP 6.3+: The slider appears in the header, so with `defer` it should also be placed in the to be available early.

This is covered by `$wp_scripts->add_data( 'twentyfourteen-slider', 'group', 0 );` below. For scripts that were moved to the header in the recent changes to allow for the loading strategy, I've modified the `group` via `*add_data()`

**Comment by felixarntz** [comment]
@peterwilsoncc 
> There's a significant difference here: in 6.3+ there is a type check for the `$args` parameter that doesn't exist for the `$in_footer` argument in earlier versions of WP.

That check is there because handling an array has more requirements. For example we can't access `$args['something']` on a boolean, so the check is needed there. That is not the case for the reverse - pretty much anything in PHP can be evaluated to a boolean. Additionally, if an array is passed to the function <6.3, that array still won't be set/stored anywhere, since it'll only interpret it as boolean and then store the intended `group: 1` value.

> A part of the motivation with these changes is to encourage future additions to these themes to use the legacy signature when supporting new features in WP (most likely the block editor). `[ `in_footer` => false ]` will load scripts in the header in older versions of WP.

Do you mean footer? `[ 'in_footer' => false ]` will load scripts in the footer in older versions, and that's why it should only be used if we want a script to be in the head in 6.3+ and in the footer in <6.3. Otherwise (i.e. if we want the script to be in the head across all versions), passing `[ 'in_footer' => false ]` is a breaking change. For exactly that reason, there is no such usage here (except for the one in https://core.trac.wordpress.org/changeset/56556/trunk/src/wp-content/themes/twentyeleven/showcase.php, which per the above is intentional)., and there is no breaking change.

Again, I'm not sure which problem you're trying to solve here. It doesn't make sense to me to encourage the legacy signature of the function when it's just that, the legacy signature, particularly as using the new signature doesn't cause backward compatibility breaks.

**Comment by peterwilsoncc** [comment]
@felixarntz The legacy signature is the only signature that is officially supported in the versions of WP that these themes need to support. For themes that only support WP 6.3 or later, then you are right, it's best not to encourage use of the legacy signature.

However, for themes that are required to support older versions of WordPress then using the updated signature can and has led to mistakes so these themes should be encouraged to use the old signature.

`[ 'in_footer' => false ]`, `[ 'strategy' => 'whatever' ]` and various other combinations can cause unintentional behaviour in earlier versions of WP and we need to account for that.

What I am trying to support in this PR is simple: making sure developers don't unintentionally introduce errors to older themes while using older methods to set data on scripts to take advantage of the new features. It doesn't revert the changes of the previous commits, it makes sure they're used in a fashion that won't introduce mistakes.

**Comment by felixarntz** [comment]
@peterwilsoncc Yes, using the new signature incorrectly can lead to errors. But that applies to any code, we can't prevent developers from making mistakes.

The way the code currently is there is no backward compatibility breakage, and I think that's a reasonable goal. I don't think it's a good idea to encourage using the legacy signature when it's the legacy signature. We should encourage the new signature, otherwise what is that even good for?

If we want to provide a way to more intuitively avoid mistakes from using the new signature in older WordPress versions, we should introduce a wrapper function that takes care of the one quirk, as it is [outlined in the dev note here](https://make.wordpress.org/core/2023/07/14/registering-scripts-with-async-and-defer-attributes-in-wordpress-6-3/#:~:text=An%20example%20of%20such%20a%20script%20registration/enqueue%20wrapper%20may%20look%20something%20as%20follows%3A).

I personally think that is overkill since the usage in the themes is not problematic as it stands, and the wrapper function would need to be duplicated across every default theme. That said, if you feel strongly about avoiding mistakes with the new signature on older WP versions, then that is an approach I would be on board with. It encourages the new signature while taking care of the one quirk that people would otherwise have to keep in mind.

**Comment by peterwilsoncc** [comment]
@felixarntz I don't think it's overkill.

The reason we are having this discussion is that the signature was changed in such a way that it's possible a developers intent is reversed when using the legacy signature. 

There were multiple comments on the initial ticket requesting the use of `add_data`, as is the case for conditional comments, but that's not how the change was implemented. The logic was to avoid the bugs that have already being introduced & since fixed.

> But that applies to any code, we can't prevent developers from making mistakes.

This is true but we can make it easier for developers. to catch mistakes.

I really don't see why it is controversial for themes that are required to support older versions of WordPress to use function signatures those versions of WordPress support. I am trying to prevent the introduction of bugs while improving performance, I don't see the problem.

**Comment by felixarntz** [comment]
Thanks @peterwilsoncc. As mentioned, I'm not opposed to introducing a wrapper function in those themes so that this kind of mistake can be avoided.

> I really don't see why it is controversial for themes that are required to support older versions of WordPress to use function signatures those versions of WordPress support.

What to me is controversial is the idea that the default themes would encourage using the old signature of the function. We can avoid that by using a wrapper function that allows using the new signature without the risk of making the mistake of having `array( 'in_footer' => false )` being interpreted as `true` in older versions.

**Comment by westonruter** [comment]
> I really don't see why it is controversial for themes that are required to support older versions of WordPress to use function signatures those versions of WordPress support. I am trying to prevent the introduction of bugs while improving performance, I don't see the problem.

@peterwilsoncc Since these themes are maintained with WordPress Core, the risk seems to be that core committers would be unaware of the new argument to `wp_enqueue_script()`/`wp_register_script()`, right? In any case, old core themes already are lacking maintenance and even more so it seems highly unlikely that any new scripts would be introduced in old core themes which would cause the problem that you've identified. So it seems to me that the risk of there being a problem is very low.

**Comment by peterwilsoncc** [comment]
> What to me is controversial is the idea that the default themes would encourage using the old signature of the function. 

I think this misrepresents what this pull request does. It encourages the use of the supported signature in the versions of WP a theme is required to support. It does this because the type casting can result in different results when using the old and new signature.

For the new features this pull request encourages their use in a forward compatible manner in themes required to support older versions of WP.

Honestly, I don't think I care enough about this to continue the discussion. I've tried to pre-empt the bugs I think it will introduce in themes required to support older versions of WP (one if which has already needed to be fixed) but given the discussion they're not my bugs to fix.
